sys/net/routing/olsr2/olsr_init.c

```diff
@@ -0,0 +1,228 @@
+        DEBUG("received %d bytes from %s", recsize, netaddr_to_str_s(&nbuf[0], &_src));
+
+        mutex_lock(&olsr_data);
+        reader_handle_packet(&buffer, recsize, &_src, 1); // TODO: proper metric
+        mutex_unlock(&olsr_data);
+    }
+
+    return NULL;
+}
+
+static void* olsr_sender_thread(void *ctx __attribute__((unused)))
+{
+    DEBUG("olsr_sender_thread, pid %d\n", thread_getpid());
+
+    /* message queue, so messages don't get lost */
+    msg_t msgq[2];
+    msg_init_queue(msgq, sizeof msgq);
+
+    while (1) {
+        msg_t m;
+        msg_receive(&m);
+        struct timer_msg *tmsg = (struct timer_msg *) m.content.ptr;
+
+        mutex_lock(&olsr_data);
+        tmsg->func();
+        mutex_unlock(&olsr_data);
+
+        /* add jitter */
+        tmsg->interval.microseconds = genrand_uint32() % OLSR2_MAX_JITTER_US;
+
+        if (vtimer_set_msg(&tmsg->timer, tmsg->interval, thread_getpid(), tmsg) != 0) {
+            DEBUG("vtimer_set_msg failed, stopped sending");
+        }
+    }
+    return NULL;
+}
+
+static ipv6_addr_t *get_next_hop(ipv6_addr_t *dest)
+{
+    struct olsr_node *node = get_node((struct netaddr *) dest); // get_node will only look at the first few bytes
+
+    if (node == NULL) {
+        return NULL;
+    }
+
+    return (ipv6_addr_t *) node->next_addr;
+}
+
+#ifdef ENABLE_NAME
+ipv6_addr_t *get_ip_by_name(char *name)
+{
+    struct olsr_node *node;
+    avl_for_each_element(get_olsr_head(), node, node) {
+        if (node->name != NULL && strcmp(node->name, name) == 0) {
+            return (ipv6_addr_t *) node->addr;
+        }
+    }
+
+    return NULL;
+}
+#endif
+
+void olsr_init(void)
+{
+
+#ifdef ENABLE_NAME
+#ifdef BOARD_NATIVE
+    olsr2_local_name = gen_name(_name, sizeof _name);
+#else
+    olsr2_local_name = sysconfig.name;
+#endif
+#endif
+    mutex_init(&olsr_data);
+    node_init();
+    reader_init();
+    writer_init(write_packet);
+
+    /* we always send to the same broadcast address, prepare it once */
+    sa_bcast.sin6_family = AF_INET6;
+    sa_bcast.sin6_port = HTONS(MANET_PORT);
+    ipv6_addr_set_all_nodes_addr(&sa_bcast.sin6_addr);
+
+    /* enable receive */
+    sock = socket_base_socket(PF_INET6, SOCK_DGRAM, IPPROTO_UDP);
+    thread_create(receive_thread_stack, sizeof receive_thread_stack, PRIORITY_MAIN - 1, CREATE_STACKTEST, olsr_receiver_thread, NULL, "olsr_rec");
+
+    /* set get_local_addr() */
+    get_local_addr()->_type = AF_INET6;
+    get_local_addr()->_prefix_len = 128;
+    ipv6_net_if_get_best_src_addr((ipv6_addr_t *) get_local_addr(), &sa_bcast.sin6_addr);
+
+    /* register olsr for routing */
+    ipv6_iface_set_routing_provider(get_next_hop);
+
+    DEBUG("This is node %s with IP %s", olsr2_local_name, netaddr_to_str_s(&nbuf[0], get_local_addr()));
+
+    /* enable sending */
+    int pid = thread_create(sender_thread_stack, sizeof sender_thread_stack, PRIORITY_MAIN - 1, CREATE_STACKTEST, olsr_sender_thread, NULL, "olsr_snd");
```

Comment:
Is it possible to combine both threads (l202[receiving] + l215[sending]) into only 1 thread?